man/rzxdump.1
.\" -*- nroff -*-
.\"
.\" rzxdump.1: rzxdump man page
.\" Copyright (C) 2002-2004 Philip Kendall
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: philip-fuse@shadowmagic.org.uk
.\"
.\"
.TH rzxdump 1 "24th May, 2013" "Version 1.1.1" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
rzxdump \(em List contents of Sinclair ZX Spectrum input recording files
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B rzxdump
.IR "rzxfile" ...
.P
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
rzxdump outputs a description of the contents of RZX input recording
files.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.I rzxfile
specifies an input recording file whose contents are to be listed.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
The contents of compressed blocks are not displayed with in a useful
fashion.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR rzxcheck "(1),"
.IR rzxtool "(1)"
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Philip Kendall (philip\-fuse@shadowmagic.org.uk).


ChangeLog
2013-05-xx  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.1 released.

        * Decreased libtool version to allow binaries built against
          libspectrum 1.0.0 to run with libspectrum 1.1.1.

2013-05-18  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.0 released.

        * Support for Spectranet in snapshots (Philip Kendall).

        * Support for Cheetah SpecDrum in snapshots (Fredrick Meunier).

        * Support reading PZX files (Fredrick Meunier).

        * Support for the TZX set signal level block (Fredrick Meunier).

        * Various minor bug fixes/improvements:
          * Add support for detecting POK files (Sergio Baldoví).
          * Advertise support for bzip2 compressed and wav files (Sergio
            Baldoví).
          * Drop support for GLib version 1.x (Alberto Garcia).
          * Add support for audiofile 0.3.x (Alberto Garcia).
          * Load all snapshot RAM pages for machines with more than 128k 
            (Fredrick Meunier; thanks, ketmar).
          * Support more SZX 1.5 features (Fredrick Meunier).
          * Modernise autoconf support (Sergio Baldoví).

2010-12-16  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.0.0 released.

        * New end of tape flag (Fredrick Meunier).

        * AF and AF' were swapped when reading and writing SZX files
          (Fredrick Meunier; thanks, velesoft).

        * Support for Fuller Box and Melodik (and similar AY interfaces) in
          snapshots (Fredrick Meunier).

        * Support for Opus Discovery in snapshots (Fredrick Meunier, Gergely
          Szasz).

        * Support for the NTSC 48K Spectrum (Philip Kendall).

        * Support for stereo sound files (Fredrick Meunier; thanks, Gilberto
          Almeida).

        * Add proper support for +2A/+3, Pentagon and Scorpion in .z80 files
          (Philip Kendall; thanks, Nick Corey)

        * Various fixes for compiling with Visual Studio 6 (Simon Owen).

        * Various minor bug fixes/improvements:
          * Looser signature checks for .dsk files (Gergely Szasz; thanks,
            zx81 and Simon Owen).
          * --without-(bzip2|zlib) options to configure for Gentoo and similar
            distributions (José Manuel Ferrer Ortiz).
          * Handle querying the block length for turbo and pure data blocks
            with 0 length (Fredrick Meunier).
          * Fix loading of Warajevo TAPs with raw samples (Fredrick Meunier).
          * Fail sensibly if file decompression fails (Fredrick Meunier).
          * Remove various unused variables, initialisations etc (Fredrick
            Meunier).

2008-11-30  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 0.5.0 released.

	* Memory management change: memory allocation functions now abort
	  on failure. Leads to simpler code both internally and for
	  users of the library (Philip Kendall).

	* Support for tape acceleration (Philip Kendall)

	* IDE improvements to allow DivIDE to support FATware (Philip
	  Kendall; thanks, http://www.dusky.sk/zxs/)

	* New API to deal with flagging automatic snapshots in RZX files
	  (Philip Kendall).

	* New function to get length (time) of a tape block (Philip Kendall,
	  Fredrick Meunier).

	* 128K SNA files are now treated as being for the Pentagon 128
	  rather than the Spectrum 128 (Fredrick Meunier)

	* SNA files now written out correctly (Philip Kendall).

	* Handle MDR write protection flag correctly (Philip Kendall)

	* Allow MDR images with bad checksums to load (the error will
	  be handled by the If1 ROM) (Gergely Szasz)

        * Support saving custom ROMs in SZX snapshots (Fredrick Meunier,
	  Stuart Brady)

	* Support for DivIDE, Kempston mouse and the simple IDE interface
	  in SZX snapshots (Fredrick Meunier)

        * Allow building without libgcrypt (Stuart Brady)

	* Amiga and MorphOS compilation fixes (Chris Young, Q-Master)

	* Don't produce empty libraries to fix build process on
	  OpenSolaris (Fredrick Meunier)

	* Remove deprecated functions (Fredrick Meunier).

	* Tweak timings for TS2068 (Fredrick Meunier; thanks Philip Kendall).

        * NTSC Spectrum support (Philip Kendall; thanks, Claudio Bernet).

2008-01-05  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.4.0 released.

	* Add support for loading from WAV files (Fredrick Meunier).

	* Add support for CSW tape files (Fredrick Meunier).

	* Tweak timings for TC2048 and TC2068 (Fredrick Meunier; thanks Philip
	  Kendall and Mark Woodmass).

	* Add support for the +D in z80 and szx snapshots (Fredrick Meunier).

	* Add Pentagon 512/1024 support (Q-Master).

	* Allow Amiga version to decompress files with xfdmaster.library
	  (Chris Young).

	* Allow MorphOS version to use xfdmaster.library (Q-Master).

	* Various other bugfixes.

2007-05-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0.1 released (bug fix release).

	* Avoid crash when reading an invalid TZX file (thanks, crabfists).
	  (Philip Kendall).

	* Export the TZX Generalized Data Block functions when building a
	  Win32 DLL (Stuart Brady).

	* Distribute the hacking/ directory (thanks, Stuart Brady) (Philip
	  Kendall).

	* Remove warning when building from CVS (Stuart Brady).

2007-04-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0 released.

	* Support for RZX rollback, including having snapshots and
	  multiple input recording blocks in input recording files (Philip
	  Kendall).

	* Improved IDE emulation (Matthew Westcott).

	* Support for microdrive images (Philip Kendall, Gergely Szasz).

	* TZX Generalized Data Block support. (Philip Kendall).

	* Support for Z80Em and CSW tape files (Darren Salt).

	* Support for the TS2068 (Fredrick Meunier).

	* Support for connected joystick information in snapshots (Fredrick
	  Meunier).

	* Add interrupt length to timing information (Philip Kendall).


rzxdump.c
/* rzxdump.c: get info on an RZX file
   Copyright (c) 2002-2003 Philip Kendall

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   E-mail: philip-fuse@shadowmagic.org.uk

*/

#include <config.h>

#include <stddef.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#ifdef HAVE_LIBGCRYPT
#include <gcrypt.h>
#endif				/* #ifdef HAVE_LIBGCRYPT */

#include <libspectrum.h>

#include "utils.h"

char *progname;

const char *rzx_signature = "RZX!";

libspectrum_word read_word( unsigned char **ptr );
libspectrum_dword read_dword( unsigned char **ptr );

static int
do_file( const char *filename );

static int
read_creator_block( unsigned char **ptr, unsigned char *end );
static int
read_snapshot_block( unsigned char **ptr, unsigned char *end );
static int
read_input_block( unsigned char **ptr, unsigned char *end );
static int read_sign_start_block( unsigned char **ptr, unsigned char *end );
static int read_sign_end_block( unsigned char **ptr, unsigned char *end );

libspectrum_word
read_word( unsigned char **ptr )
{
  libspectrum_word result = (*ptr)[0] + (*ptr)[1] * 0x100;
  (*ptr) += 2;
  return result;
}

libspectrum_dword
read_dword( unsigned char **ptr )
{
  libspectrum_dword result = (*ptr)[0]             +
                             (*ptr)[1] *     0x100 +
                             (*ptr)[2] *   0x10000 +
                             (*ptr)[3] * 0x1000000;
  (*ptr) += 4;
  return result;
}


int main( int argc, char **argv )
{
  int i;

  progname = argv[0];

  if( argc < 2 ) {
    fprintf( stderr, "%s: Usage: %s <rzx file(s)>\n", progname, progname );
    return 1;
  }

  for( i=1; i<argc; i++ ) {
    do_file( argv[i] );
  }    

  return 0;
}

static int
do_file( const char *filename )
{
  unsigned char *buffer, *ptr, *end; size_t length;
  int error;

  printf( "Examining file %s\n", filename );

  error = read_file( filename, &buffer, &length ); if( error ) return error;

  ptr = buffer; end = buffer + length;

  /* Read the RZX header */

  if( end - ptr < (ptrdiff_t)strlen( rzx_signature ) + 6 ) {
    fprintf( stderr,
	     "%s: Not enough bytes for RZX header (%ld bytes)\n",
	     progname, (unsigned long)strlen( rzx_signature ) + 6 );
    free( buffer );
    return 1;
  }

  if( memcmp( ptr, rzx_signature, strlen( rzx_signature ) ) ) {
    fprintf( stderr, "%s: Wrong signature: expected `%s'\n", progname,
	     rzx_signature );
    free( buffer );
    return 1;
  }

  printf( "Found RZX signature\n" );
  ptr += strlen( rzx_signature );

  printf( "  Major version: %d\n", (int)*ptr++ );
  printf( "  Minor version: %d\n", (int)*ptr++ );
  printf( "  Flags: %d\n", read_dword( &ptr ) );

  while( ptr < end ) {

    unsigned char id;

    id = *ptr++;

    switch( id ) {

    case 0x10: error = read_creator_block( &ptr, end ); break; 
    case 0x20: error = read_sign_start_block( &ptr, end ); break;
    case 0x21: error = read_sign_end_block( &ptr, end ); break;
    case 0x30: error = read_snapshot_block( &ptr, end ); break;
    case 0x80: error = read_input_block( &ptr, end ); break;

    default:
      fprintf( stderr, "%s: Unknown block type 0x%02x\n", progname, id );
      free( buffer );
      return 1;

    }

    if( error ) { free( buffer ); return 1; }
  }

  free( buffer );

  return 0;
}

static int
read_creator_block( unsigned char **ptr, unsigned char *end )
{
  size_t length;

  if( end - *ptr < 28 ) {
    fprintf( stderr,
	     "%s: Not enough bytes for creator block\n", progname );
    return 1;
  }

  printf( "Found a creator block\n" );

  length = read_dword( ptr );
  printf( "  Length: %ld bytes\n", (unsigned long)length );
  printf( "  Creator: `%s'\n", *ptr ); (*ptr) += 20;
  printf( "  Creator major version: %d\n", read_word( ptr ) );
  printf( "  Creator minor version: %d\n", read_word( ptr ) );
  printf( "  Creator custom data: %ld bytes\n", (unsigned long)length - 29 );
  (*ptr) += length - 29;

  return 0;
}

static int
read_snapshot_block( unsigned char **ptr, unsigned char *end )
{
  size_t block_length;
  libspectrum_dword flags;

  if( end - *ptr < 16 ) {
    fprintf( stderr,
	     "%s: Not enough bytes for snapshot block\n", progname );
    return 1;
  }

  printf( "Found a snapshot block\n" );

  block_length = read_dword( ptr );
  printf( "  Length: %lu bytes\n", (unsigned long)block_length );

  flags = read_dword( ptr );
  printf( "  Flags: %lu\n", (unsigned long)flags );
  printf( "  Snapshot extension: `%s'\n", *ptr ); (*ptr) += 4;
  printf( "  Snap length: %d bytes\n", read_dword( ptr ) );

  if( ( flags & 0x01 ) && !( flags & 0x02 ) )
    printf( "  Snap path: %s\n", *ptr + 4 );

  (*ptr) += block_length - 17;

  return 0;
}

static int
read_input_block( unsigned char **ptr, unsigned char *end )
{
  size_t i, frames, length; int flags;

  if( end - *ptr < 17 ) {
    fprintf( stderr,
	     "%s: Not enough bytes for input recording block\n", progname );
    return 1;
  }

  printf( "Found an input recording block\n" );
  
  length = read_dword( ptr );
  printf( "  Length: %ld bytes\n", (unsigned long)length );

  frames = read_dword( ptr );
  printf( "  Frame count: %ld\n", (unsigned long)frames );

  printf( "  Frame length (obsolete): %d bytes\n", *(*ptr)++ );
  printf( "  Tstate counter: %d\n", read_dword( ptr ) );

  flags = read_dword( ptr );
  printf( "  Flags: %d\n", flags );

  /* Check there's enough data */
  if( end - *ptr < (off_t)length - 18 ) {
    fprintf( stderr, "%s: not enough frame data\n", progname );
    return 1;
  }

  if( flags & 0x02 ) {		/* Data is compressed */
    printf( "  Skipping compressed data\n" );
    (*ptr) += length - 18;

  } else {			/* Data is not compressed */

    for( i=0; i<frames; i++ ) {

      size_t count;

      printf( "Examining frame %ld\n", (unsigned long)i );
    
      if( end - *ptr < 4 ) {
	fprintf( stderr, "%s: Not enough data for frame %ld\n", progname,
		 (unsigned long)i );
	return 1;
      }

      printf( "  Instruction count: %d\n", read_word( ptr ) );

      count = read_word( ptr );
      if( count == 0xffff ) {
	printf( "  (Repeat last frame's INs)\n" );
	continue;
      }

      printf( "  IN count: %ld\n", (unsigned long)count );

      if( end - *ptr < (ptrdiff_t)count ) {
	fprintf( stderr,
		 "%s: Not enough data for frame %ld (expected %ld bytes)\n",
		 progname, (unsigned long)i, (unsigned long)count );
	return 1;
      }

      (*ptr) += count;
    }

  }

  return 0;
}

static int
read_sign_start_block( unsigned char **ptr, unsigned char *end )
{
  size_t length;

  if( end - *ptr < 8 ) {
    fprintf( stderr, "%s: not enough bytes for sign start block\n", progname );
    return 1;
  }

  printf( "Found a signed data start block\n" );

  length = read_dword( ptr );

  printf( "  Length: %ld bytes\n", (unsigned long)length );
  printf( "  Key ID: 0x%08x\n", (unsigned int)read_dword( ptr ) );
  printf( "  Week code: 0x%08x\n", (unsigned int)read_dword( ptr ) );

  (*ptr) += length - 13;

  return 0;
}

static int
read_sign_end_block( unsigned char **ptr, unsigned char *end )
{
  size_t length;
#ifdef HAVE_LIBGCRYPT
  gcry_mpi_t a; int error; size_t length2;
  unsigned char *buffer;
#endif				/* #ifdef HAVE_LIBGCRYPT */

  if( end - *ptr < 4 ) {
    fprintf( stderr, "%s: not enough bytes for sign end block\n", progname );
    return 1;
  }

  printf( "Found a signed data end block\n" );

  length = read_dword( ptr );
  printf( "  Length: %ld bytes\n", (unsigned long)length );

  length -= 5;

  if( end - *ptr < (ptrdiff_t)length ) {
    fprintf( stderr, "%s: not enough bytes for sign end block\n", progname );
    return 1;
  }

#ifdef HAVE_LIBGCRYPT

  error = gcry_mpi_scan( &a, GCRYMPI_FMT_PGP, *ptr, length, &length2 );
  if( error ) {
    fprintf( stderr, "%s: error reading r: %s\n", progname,
	     gcry_strerror( error ) );
    return 1;
  }
  *ptr += length2; length -= length2;

  error = gcry_mpi_aprint( GCRYMPI_FMT_HEX, &buffer, NULL, a );
  printf( "  r: %s\n", buffer );
  free( buffer ); gcry_mpi_release( a );

  error = gcry_mpi_scan( &a, GCRYMPI_FMT_PGP, *ptr, length, &length2 );
  if( error ) {
    fprintf( stderr, "%s: error reading s: %s\n", progname,
	     gcry_strerror( error ) );
    return 1;
  }
  *ptr += length2; length -= length2;

  error = gcry_mpi_aprint( GCRYMPI_FMT_HEX, &buffer, NULL, a );
  printf( "  s: %s\n", buffer );
  free( buffer ); gcry_mpi_release( a );

#endif

  (*ptr) += length;

  return 0;
}
